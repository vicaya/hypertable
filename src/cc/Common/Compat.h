/**
 * Copyright (C) 2007 Luke Lu (Zvents, Inc.)
 *
 * This file is part of Hypertable.
 *
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or any later version.
 *
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

#ifndef HYPERTABLE_COMPAT_H
#define HYPERTABLE_COMPAT_H

// Include this before other includes .cc files

// The same stuff for C code
#include "compat-c.h"

#include <cstddef> // for std::size_t and std::ptrdiff_t

// C++ specific stuff
#ifndef BOOST_SPIRIT_THREADSAFE
#  define BOOST_SPIRIT_THREADSAFE
#endif

#define BOOST_IOSTREAMS_USE_DEPRECATED

#include "Sweetener.h"

#define HT_UNUSED(x) static_cast<void>(x)

#endif // HYPERTABLE_COMPAT_H
