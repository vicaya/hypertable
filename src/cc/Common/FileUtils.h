/**
 * Copyright (C) 2007 Doug Judd (Zvents, Inc.)
 *
 * This file is part of Hypertable.
 *
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or any later version.
 *
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */
#ifndef HYPERTABLE_FDUTILS_H
#define HYPERTABLE_FDUTILS_H

extern "C" {
#include <sys/socket.h>
#include <sys/types.h>
}
#include "Common/String.h"

namespace Hypertable {

  class FileUtils {

  public:
    static ssize_t read(const String &fname, String &contents);
    static ssize_t read(int fd, void *vptr, size_t n);
    static ssize_t pread(int fd, void *vptr, size_t n, off_t offset);
    static ssize_t write(const String &fname, String &contents);
    static ssize_t write(int fd, const void *vptr, size_t n);
    static ssize_t writev(int fd, const struct iovec *vector, int count);
    static ssize_t sendto(int fd, const void *vptr, size_t n,
                          const sockaddr *to, socklen_t tolen);
    static ssize_t send(int fd, const void *vptr, size_t n);
    static ssize_t recvfrom(int fd, void *vptr, size_t n,
                            struct sockaddr *from, socklen_t *fromlen);
    static ssize_t recv(int fd, void *vptr, size_t n);
    static void set_flags(int fd, int flags);
    static char *file_to_buffer(const String &fname, off_t *lenp);
    static String file_to_string(const String &fname);
    static bool mkdirs(const String &dirname);
    static bool exists(const String &fname);
    static bool unlink(const String &fname);
    static bool rename(const String &oldpath, const String &newpath);
    static uint64_t size(const String &fname);
    static off_t length(const String &fname);
    static void add_trailing_slash(String &path);
    static bool expand_tilde(String &fname);
#ifdef HT_XATTR_ENABLED
    static int
    getxattr(const String &path, const String &name, void *value, size_t size);
    static int fgetxattr(int fd, const String &name, void *value, size_t size);
    static int
    setxattr(const String &path, const String &name, const void *value,
             size_t size, int flags);
    static int
    fsetxattr(int fd, const String &name, const void *value, size_t size,
              int flags);
    static int removexattr(const String &path, const String &name);
    static int fremovexattr(int fd, const String &name);
#endif
  };

}

#endif // HYPERTABLE_FDUITLS_H

