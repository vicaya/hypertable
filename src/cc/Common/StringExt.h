/**
 * Copyright (C) 2008 Doug Judd (Zvents, Inc.)
 *
 * This file is part of Hypertable.
 *
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or any later version.
 *
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */
#ifndef HYPERTABLE_STRINGEXT_H
#define HYPERTABLE_STRINGEXT_H

#include <cstdio>
#include <stdexcept>
#include <set>

#include "Common/String.h"
#include "HashMap.h"


typedef std::set<std::string> StringSet;


/** STL Strict Weak Ordering for comparing c-style strings. */
struct LtCstr {
  bool operator()(const char* s1, const char* s2) const {
    return strcmp(s1, s2) < 0;
  }
};

typedef std::set<const char *, LtCstr>  CstrSet;

inline std::string operator+(const std::string& s1, short sval) {
  char cbuf[8];
  sprintf(cbuf, "%d", sval);
  return s1 + cbuf;
}

inline std::string operator+(const std::string& s1, int ival) {
  char cbuf[16];
  sprintf(cbuf, "%d", ival);
  return s1 + cbuf;
}

inline std::string operator+(const std::string& s1, uint32_t ival) {
  char cbuf[16];
  sprintf(cbuf, "%d", ival);
  return s1 + cbuf;
}

inline std::string operator+(const std::string& s1, int64_t llval) {
  char cbuf[32];
  sprintf(cbuf, "%lld", (long long int)llval);
  return s1 + cbuf;
}

inline std::string operator+(const std::string& s1, uint64_t llval) {
  char cbuf[32];
  sprintf(cbuf, "%llu", (long long unsigned int)llval);
  return s1 + cbuf;
}


#endif // HYPERTABLE_STRINGEXT_H
