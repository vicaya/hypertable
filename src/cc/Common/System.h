/**
 * Copyright (C) 2007 Doug Judd (Zvents, Inc.)
 *
 * This file is part of Hypertable.
 *
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or any later version.
 *
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

#ifndef HYPERTABLE_SYSTEM_H
#define HYPERTABLE_SYSTEM_H

#include <boost/random.hpp>
#include "Common/Version.h"
#include "Common/Mutex.h"
#include "Common/String.h"

namespace Hypertable {

  class CpuInfo;
  class CpuStat;
  class LoadAvgStat;
  class MemStat;
  class DiskStat;
  class OsInfo;
  class SwapStat;
  class NetInfo;
  class NetStat;
  class ProcInfo;
  class ProcStat;
  class FsStat;
  class TermInfo;

  class System {
  public:
    // must be inlined to do proper version check
    static inline void initialize(const String &install_directory = String()) {
      ScopedLock lock(ms_mutex);

      if (ms_initialized)
        return;

      check_version();
      _init(install_directory);
      ms_initialized = true;
    }

    static String locate_install_dir(const char *argv0);
    static String _locate_install_dir(const char *argv0);

    static String install_dir;
    static String exe_name;

    static int32_t get_processor_count();
    static int32_t get_pid();

    static uint32_t rand32() { return ms_rng(); }
    static uint64_t rand64() { return (uint64_t)rand32() << 32 | rand32(); }

    // system info objects
    static const CpuInfo &cpu_info();
    static const CpuStat &cpu_stat();
    static const MemStat &mem_stat();
    static const DiskStat &disk_stat();
    static const OsInfo &os_info();
    static const SwapStat &swap_stat();
    static const NetInfo &net_info();
    static const NetStat &net_stat();
    static const ProcInfo &proc_info();
    static const ProcStat &proc_stat();
    static const FsStat &fs_stat();
    static const TermInfo &term_info();
    static const LoadAvgStat &loadavg_stat();

  private:
    static void _init(const String &install_directory);

    static bool ms_initialized;
    static Mutex ms_mutex;
    static boost::mt19937 ms_rng;
  };

} // namespace Hypertable

#endif // HYPERTABLE_SYSTEM_H
