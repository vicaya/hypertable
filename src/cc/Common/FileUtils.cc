/**
 * Copyright (C) 2008 Doug Judd (Zvents, Inc.)
 *
 * This file is part of Hypertable.
 *
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or any later version.
 *
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

#include "Common/Compat.h"

#include <iomanip>
#include <iostream>
#include <sstream>

extern "C" {
#include <unistd.h>
#include <errno.h>
#include <pwd.h>
#include <fcntl.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <sys/uio.h>
#ifdef HT_XATTR_ENABLED
# if defined(__FreeBSD__)
#   include <sys/extattr.h>
# else
#   include <sys/xattr.h>
# endif
# if defined(__linux__)
#   include <attr/xattr.h>
# endif
#endif
}

#include <boost/shared_array.hpp>

#include "FileUtils.h"
#include "Logger.h"

using namespace Hypertable;
using namespace std;

ssize_t FileUtils::read(const String &fname, String &contents) {
  off_t len = 0;
  String str;
  char *buf = file_to_buffer(fname, &len);
  if (buf != 0) {
    contents.append(buf, len);
    delete [] buf;
  }
  return (ssize_t)len;
}



/**
 */
ssize_t FileUtils::read(int fd, void *vptr, size_t n) {
  size_t nleft;
  ssize_t nread;
  char *ptr;

  ptr = (char *)vptr;
  nleft = n;
  while (nleft > 0) {
    if ((nread = ::read(fd, ptr, nleft)) < 0) {
      if (errno == EINTR)
        nread = 0;/* and call read() again */
      else if (errno == EAGAIN)
        break;
      else {
        return -1;
      }
    } else if (nread == 0)
      break;/* EOF */

    nleft -= nread;
    ptr   += nread;
  }
  return n - nleft;
}

/**
 */
ssize_t FileUtils::pread(int fd, void *vptr, size_t n, off_t offset) {
  size_t nleft;
  ssize_t nread;
  char *ptr;

  ptr = (char *)vptr;
  nleft = n;
  while (nleft > 0) {
    if ((nread = ::pread(fd, ptr, nleft, offset)) < 0) {
      if (errno == EINTR)
        nread = 0;/* and call read() again */
      else if (errno == EAGAIN)
        break;
      else {
        return -1;
      }
    } else if (nread == 0)
      break;/* EOF */

    nleft -= nread;
    ptr   += nread;
    offset += nread;
  }
  return n - nleft;
}


ssize_t FileUtils::write(const String &fname, String &contents) {
  int fd = open(fname.c_str(), O_WRONLY|O_CREAT|O_TRUNC, 0644);
  if (fd < 0) {
    int saved_errno = errno;
    HT_ERRORF("Unable to open file \"%s\" for writing - %s", fname.c_str(),
              strerror(saved_errno));
    errno = saved_errno;
    return -1;
  }
  ssize_t rval = write(fd, contents.c_str(), contents.length());
  ::close(fd);
  return rval;
}



/**
 */
ssize_t FileUtils::write(int fd, const void *vptr, size_t n) {
  size_t nleft;
  ssize_t nwritten;
  const char *ptr;

  ptr = (const char *)vptr;
  nleft = n;
  while (nleft > 0) {
    if ((nwritten = ::write(fd, ptr, nleft)) <= 0) {
      if (errno == EINTR)
        nwritten = 0; /* and call write() again */
      else if (errno == EAGAIN)
        break;
      else {
        return -1; /* error */
      }
    }

    nleft -= nwritten;
    ptr   += nwritten;
  }
  return n - nleft;
}

ssize_t FileUtils::writev(int fd, const struct iovec *vector, int count) {
  ssize_t nwritten;
  while ((nwritten = ::writev(fd, vector, count)) <= 0) {
    if (errno == EINTR)
      nwritten = 0; /* and call write() again */
    else if (errno == EAGAIN) {
      nwritten = 0;
      break;
    }
    else {
      return -1; /* error */
    }
  }
  return nwritten;
}


ssize_t
FileUtils::sendto(int fd, const void *vptr, size_t n, const sockaddr *to,
                  socklen_t tolen) {
  size_t nleft;
  ssize_t nsent;
  const char *ptr;

  ptr = (const char *)vptr;
  nleft = n;
  while (nleft > 0) {
    if ((nsent = ::sendto(fd, ptr, nleft, 0, to, tolen)) <= 0) {
      if (errno == EINTR)
        nsent = 0; /* and call sendto() again */
      else if (errno == EAGAIN || errno == ENOBUFS)
        break;
      else {
        return -1; /* error */
      }
    }

    nleft -= nsent;
    ptr   += nsent;
  }
  return n - nleft;
}



ssize_t FileUtils::send(int fd, const void *vptr, size_t n) {
  size_t nleft;
  ssize_t nsent;
  const char *ptr;

  ptr = (const char *)vptr;
  nleft = n;
  while (nleft > 0) {
    if ((nsent = ::send(fd, ptr, nleft, 0)) <= 0) {
      if (errno == EINTR)
        nsent = 0; /* and call sendto() again */
      else if (errno == EAGAIN || errno == ENOBUFS)
        break;
      else {
        return -1; /* error */
      }
    }

    nleft -= nsent;
    ptr   += nsent;
  }
  return n - nleft;
}



ssize_t
FileUtils::recvfrom(int fd, void *vptr, size_t n, sockaddr *from,
                    socklen_t *fromlen) {
  ssize_t nread;
  while (true) {
    if ((nread = ::recvfrom(fd, vptr, n, 0, from, fromlen)) < 0) {
      if (errno != EINTR)
        break;
    }
    else
      break;
  }
  return nread;
}


ssize_t FileUtils::recv(int fd, void *vptr, size_t n) {
  ssize_t nread;
  while (true) {
    if ((nread = ::recv(fd, vptr, n, 0)) < 0) {
      if (errno != EINTR)
        break;
    }
    else
      break;
  }
  return nread;
}


/* flags are file status flags to turn on */
void FileUtils::set_flags(int fd, int flags) {
  int val;

  if ((val = fcntl(fd, F_GETFL, 0)) < 0) {
    int saved_errno = errno;
    cerr << "fcnt(F_GETFL) failed : " << strerror(saved_errno) << endl;
    errno = saved_errno;
  }

  val |= flags;

  if (fcntl(fd, F_SETFL, val) < 0) {
    int saved_errno = errno;
    cerr << "fcnt(F_SETFL) failed : " << strerror(saved_errno) << endl;
    errno = saved_errno;
  }
}



/**
 */
char *FileUtils::file_to_buffer(const String &fname, off_t *lenp) {
  struct stat statbuf;
  int fd;

  *lenp = 0;

  if ((fd = open(fname.c_str(), O_RDONLY)) < 0) {
    int saved_errno = errno;
    HT_ERRORF("open(\"%s\") failure - %s", fname.c_str(),  strerror(saved_errno));
    errno = saved_errno;
    return 0;
  }

  if (fstat(fd, &statbuf) < 0) {
    int saved_errno = errno;
    HT_ERRORF("fstat(\"%s\") failure - %s", fname.c_str(),  strerror(saved_errno));
    errno = saved_errno;
    return 0;
  }

  *lenp = statbuf.st_size;

  char *rbuf = new char [*lenp + 1];

  ssize_t nread = FileUtils::read(fd, rbuf, *lenp);

  ::close(fd);

  if (nread == (ssize_t)-1) {
    int saved_errno = errno;
    HT_ERRORF("read(\"%s\") failure - %s", fname.c_str(),  strerror(saved_errno));
    errno = saved_errno;
    delete [] rbuf;
    *lenp = 0;
    return 0;
  }

  if (nread < *lenp) {
    HT_WARNF("short read (%d of %d bytes)", (int)nread, (int)*lenp);
    *lenp = nread;
  }

  rbuf[nread] = 0;

  return rbuf;
}

String FileUtils::file_to_string(const String &fname) {
  String str;
  off_t len;
  char *contents = file_to_buffer(fname, &len);
  str = (contents == 0) ? "" : contents;
  delete [] contents;
  return str;
}



bool FileUtils::mkdirs(const String &dirname) {
  struct stat statbuf;
  boost::shared_array<char> tmp_dir(new char [dirname.length() + 1]);
  char *tmpdir = tmp_dir.get();
  char *ptr = tmpdir+1;

  strcpy(tmpdir, dirname.c_str());

  while ((ptr = strchr(ptr, '/')) != 0) {
    *ptr = 0;
    if (stat(tmpdir, &statbuf) != 0) {
      if (errno == ENOENT) {
        if (mkdir(tmpdir, 0755) != 0) {
          int saved_errno = errno;
          HT_ERRORF("Problem creating directory '%s' - %s",
                    tmpdir, strerror(saved_errno));
          errno = saved_errno;
          return false;
        }
      }
      else {
        int saved_errno = errno;
        HT_ERRORF("Problem stat'ing directory '%s' - %s",
                  tmpdir, strerror(saved_errno));
        errno = saved_errno;
        return false;
      }
    }
    *ptr++ = '/';
  }

  if (stat(tmpdir, &statbuf) != 0) {
    if (errno == ENOENT) {
      if (mkdir(tmpdir, 0755) != 0) {
        int saved_errno = errno;
        HT_ERRORF("Problem creating directory '%s' - %s",
                  tmpdir, strerror(saved_errno));
        errno = saved_errno;
        return false;
      }
    }
    else {
      int saved_errno = errno;
      HT_ERRORF("Problem stat'ing directory '%s' - %s",
                tmpdir, strerror(saved_errno));
      errno = saved_errno;
      return false;
    }
  }

  return true;
}


bool FileUtils::exists(const String &fname) {
  struct stat statbuf;
  if (stat(fname.c_str(), &statbuf) != 0)
    return false;
  return true;
}

bool FileUtils::unlink(const String &fname) {
  if (::unlink(fname.c_str()) == -1) {
    int saved_errno = errno;
    HT_ERRORF("unlink(\"%s\") failed - %s", fname.c_str(), strerror(saved_errno));
    errno = saved_errno;
    return false;
  }
  return true;
}

bool FileUtils::rename(const String &oldpath, const String &newpath) {
  if (::rename(oldpath.c_str(), newpath.c_str()) == -1) {
    int saved_errno = errno;
    HT_ERRORF("rename(\"%s\", \"%s\") failed - %s",
              oldpath.c_str(), newpath.c_str(), strerror(saved_errno));
    errno = saved_errno;
    return false;
  }
  return true;
}

uint64_t FileUtils::size(const String &fname) {
  struct stat statbuf;
  if (stat(fname.c_str(), &statbuf) != 0)
    return 0;
  return statbuf.st_size;

}


off_t FileUtils::length(const String &fname) {
  struct stat statbuf;
  if (stat(fname.c_str(), &statbuf) != 0)
    return (off_t)-1;
  return statbuf.st_size;
}


void FileUtils::add_trailing_slash(String &path) {
  if (path.find('/', path.length()-1) == string::npos)
    path += "/";
}


bool FileUtils::expand_tilde(String &fname) {
  struct passwd pbuf;
  struct passwd *prbuf;
  char buf[256];

  if (fname[0] != '~')
    return false;

  if (fname[1] == '/') {
    if (getpwuid_r(getuid() , &pbuf, buf, 256, &prbuf) != 0 || prbuf == 0)
      return false;
    fname = (String)pbuf.pw_dir + fname.substr(1);
  }
  else {
    String name;
    size_t first_slash = fname.find_first_of('/');

    if (first_slash == string::npos)
      name = fname.substr(1);
    else
      name = fname.substr(1, first_slash-1);

    if (getpwnam_r(name.c_str() , &pbuf, buf, 256, &prbuf) != 0 || prbuf == 0)
      return false;

    if (first_slash == string::npos)
      fname = pbuf.pw_dir;
    else
      fname = (String)pbuf.pw_dir + fname.substr(first_slash);
  }

  return true;
}


#ifdef HT_XATTR_ENABLED

int
FileUtils::getxattr(const String &path, const String &name, void *value,
                    size_t size) {
  String canonic = (String)"user." + name;
#if defined(__linux__)
  return ::getxattr(path.c_str(), canonic.c_str(), value, size);
#elif defined(__APPLE__)
  return ::getxattr(path.c_str(), canonic.c_str(), value, size, 0, 0);
#elif defined(__FreeBSD__)
  return ::extattr_get_file(path.c_str(), EXTATTR_NAMESPACE_USER, canonic.c_str(), value, size);
#else
  ImplementMe;
#endif
}


int
FileUtils::setxattr(const String &path, const String &name, const void *value,
                    size_t size, int flags) {
  String canonic = (String)"user." + name;
#if defined(__linux__)
  return ::setxattr(path.c_str(), canonic.c_str(), value, size, flags);
#elif defined(__APPLE__)
  return ::setxattr(path.c_str(), canonic.c_str(), value, size, 0, flags);
#elif defined(__FreeBSD__)
  return ::extattr_set_file(path.c_str(), EXTATTR_NAMESPACE_USER, canonic.c_str(), value, size);
#else
  ImplementMe;
#endif
}


int FileUtils::fgetxattr(int fd, const String &name, void *value, size_t size) {
  String canonic = (String)"user." + name;
#if defined(__linux__)
  return ::fgetxattr(fd, canonic.c_str(), value, size);
#elif defined(__APPLE__)
  return ::fgetxattr(fd, canonic.c_str(), value, size, 0, 0);
#elif defined(__FreeBSD__)
  return ::extattr_get_fd(fd, EXTATTR_NAMESPACE_USER, canonic.c_str(), value, size);
#else
  ImplementMe;
#endif
}


int
FileUtils::fsetxattr(int fd, const String &name, const void *value,
                     size_t size, int flags) {
  String canonic = (String)"user." + name;
#if defined(__linux__)
  return ::fsetxattr(fd, canonic.c_str(), value, size, flags);
#elif defined(__APPLE__)
  return ::fsetxattr(fd, canonic.c_str(), value, size, 0, flags);
#elif defined(__FreeBSD__)
  return ::extattr_set_fd(fd, EXTATTR_NAMESPACE_USER, canonic.c_str(), value, size);
#else
  ImplementMe;
#endif
}


int FileUtils::removexattr(const String &path, const String &name) {
  String canonic = (String)"user." + name;
#if defined(__linux__)
  return ::removexattr(path.c_str(), canonic.c_str());
#elif defined(__APPLE__)
  return ::removexattr(path.c_str(), canonic.c_str(), 0);
#elif defined(__FreeBSD__)
  return ::extattr_delete_file(path.c_str(), EXTATTR_NAMESPACE_USER, canonic.c_str());
#else
  ImplementMe;
#endif
}

int FileUtils::fremovexattr(int fd, const String &name) {
  String canonic = (String)"user." + name;
#if defined(__linux__)
  return ::fremovexattr(fd, canonic.c_str());
#elif defined(__APPLE__)
  return ::fremovexattr(fd, canonic.c_str(), 0);
#elif defined(__FreeBSD__)
  return ::extattr_delete_fd(fd, EXTATTR_NAMESPACE_USER, canonic.c_str());
#else
  ImplementMe;
#endif

}

#endif // HT_XATTR_ENABLED
