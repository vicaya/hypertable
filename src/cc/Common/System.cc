/**
 * Copyright (C) 2008 Doug Judd (Zvents, Inc.)
 *
 * This file is part of Hypertable.
 *
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or any later version.
 *
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

#include "Common/Compat.h"

#include <boost/algorithm/string.hpp>
#include <boost/filesystem.hpp>

#include <iostream>

#include "Common/Logger.h"
#include "Common/Path.h"
#include "Common/SystemInfo.h"

using namespace Hypertable;
using namespace std;
using namespace boost::filesystem;

string System::install_dir;
string System::exe_name;
bool   System::ms_initialized = false;
Mutex        System::ms_mutex;
boost::mt19937 System::ms_rng;

String System::locate_install_dir(const char *argv0) {
  ScopedLock lock(ms_mutex);
  return _locate_install_dir(argv0);
}

String System::_locate_install_dir(const char *argv0) {

  if (!install_dir.empty())
    return install_dir;

  exe_name = Path(argv0).filename();

  Path exepath(proc_info().exe);

  // Detect install_dir/bin/exe_name: assumed install layout
  if (exepath.parent_path().filename() == "bin")
    install_dir = exepath.parent_path().parent_path().directory_string();
  else
    install_dir = exepath.parent_path().directory_string();

  return install_dir;
}


void System::_init(const String &install_directory) {
  // seed the random number generator
  ms_rng.seed((uint32_t)getpid());

  if (install_directory.empty()) {
    install_dir = _locate_install_dir(proc_info().exe.c_str());
  }
  else {
    // set installation directory
    install_dir = install_directory;
    while (boost::ends_with(install_dir, "/"))
    install_dir = install_dir.substr(0, install_dir.length()-1);
  }

  if (exe_name.empty())
    exe_name = Path(proc_info().args[0]).filename();

  // initialize logging system
  Logger::initialize(exe_name);
}


int32_t System::get_processor_count() {
  return cpu_info().total_cores;
}

int32_t System::get_pid() {
  return proc_info().pid;
}
