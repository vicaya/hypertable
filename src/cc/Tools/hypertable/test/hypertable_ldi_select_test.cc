/**
 * Copyright (C) 2010 Sanjit Jhala(Hypertable, Inc.)
 *
 * This file is part of Hypertable.
 *
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or any later version.
 *
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

#include "Common/Compat.h"
#include <cstdlib>
#include <iostream>
#include <string>

extern "C" {
#include <sys/types.h>
#include <unistd.h>
}

#include "Common/StringExt.h"
#include "Common/FileUtils.h"
#include "Common/Logger.h"
#include "Common/System.h"
#include "DfsBroker/Lib/Client.h"

using namespace Hypertable;
using namespace std;


namespace {
  const char *required_files[] = {
    "./hypertable_ldi_select_test",
    "./hypertable.cfg",
    "./hypertable_ldi_stdin_test_load.hql",
    "./hypertable_ldi_stdin_test_select.hql",
    "./hypertable_test.tsv.gz",
    "./hypertable_ldi_select_test.golden",
    0
  };
}


int main(int argc, char **argv) {
  String cmd_str;

  if (argc != 2) {
    HT_ERRORF("Usage: %s $INSTALL_DIR", argv[0]);
    _exit(1);
  }
  String install_dir = argv[1];
  System::initialize(argv[0]);

  for (int i=0; required_files[i]; i++) {
    if (!FileUtils::exists(required_files[i])) {
      HT_ERRORF("Unable to find '%s'", required_files[i]);
      _exit(1);
    }
  }

  /**
   * LDI and Select using stdin
   */
  cmd_str = "./hypertable --test-mode --config hypertable.cfg "
      "< hypertable_ldi_stdin_test_load.hql > hypertable_ldi_select_test.output 2>&1";
  if (system(cmd_str.c_str()) != 0)
    _exit(1);

  cmd_str = "./hypertable --test-mode --config hypertable.cfg "
      "< hypertable_ldi_stdin_test_select.hql >> hypertable_ldi_select_test.output 2>&1";
  if (system(cmd_str.c_str()) != 0)
    _exit(1);

  /**
   * LDI and Select using DfsBroker
   */
  String test_dir = install_dir + "/fs/local/ldi_test/";
  // copy data file to dfs dir

  cmd_str = "rm -rf " + test_dir;
  if (system(cmd_str.c_str()) != 0)
    _exit(1);
  cmd_str = "mkdir " + test_dir;
  if (system(cmd_str.c_str()) != 0)
    _exit(1);
  cmd_str = "cp hypertable_test.tsv.gz " + test_dir;
  if (system(cmd_str.c_str()) != 0)
    _exit(1);

  String hql = (String)" USE \"/test\";" +
                " DROP TABLE IF EXISTS hypertable;" +
                " CREATE TABLE hypertable ( TestColumnFamily );" +
                " LOAD DATA INFILE ROW_KEY_COLUMN=rowkey" +
                " \"dfs:///ldi_test/hypertable_test.tsv.gz\"" +
                " INTO TABLE hypertable;"
                ;
  // load from dfs zipped file
  cmd_str = "./hypertable --test-mode --config hypertable.cfg --exec '"+ hql + "'";
  if (system(cmd_str.c_str()) != 0)
    _exit(1);
  // select into dfs zipped file
  hql = "USE \"/test\"; SELECT * FROM hypertable INTO FILE \"dfs:///ldi_test/dfs_select.gz\";";
  cmd_str = "./hypertable --test-mode --config hypertable.cfg --exec '"+ hql + "'";
  if (system(cmd_str.c_str()) != 0)
    _exit(1);

  // cp from dfs dir
  cmd_str = "cp " + test_dir + "dfs_select.gz .";
  if (system(cmd_str.c_str()) != 0)
    _exit(1);

  cmd_str = "rm -rf " + test_dir;
  if (system(cmd_str.c_str()) != 0)
    _exit(1);

  cmd_str = "gunzip -f dfs_select.gz";
  if (system(cmd_str.c_str()) != 0)
    _exit(1);

  cmd_str = "cat dfs_select >> hypertable_ldi_select_test.output ";
  if (system(cmd_str.c_str()) != 0)
    _exit(1);

  cmd_str = "diff hypertable_ldi_select_test.output hypertable_ldi_select_test.golden";
  if (system(cmd_str.c_str()) != 0)
    _exit(1);

  hql = (String)" USE \"/test\";" +
    " DROP TABLE IF EXISTS hypertable;" +
    " CREATE TABLE hypertable ( TestColumnFamily );" +
    " LOAD DATA INFILE \"hypertable_escape_test.tsv\" INTO TABLE hypertable;" +
    " SELECT * FROM hypertable NO_ESCAPE INTO FILE \"hypertable_escape_test.output\";"
    ;

  // load from dfs zipped file
  cmd_str = "./hypertable --test-mode --config hypertable.cfg --exec '"+ hql + "'";
  if (system(cmd_str.c_str()) != 0)
    _exit(1);

  cmd_str = "diff hypertable_escape_test.output hypertable_escape_test.golden";
  if (system(cmd_str.c_str()) != 0)
    _exit(1);

  _exit(0);
}
