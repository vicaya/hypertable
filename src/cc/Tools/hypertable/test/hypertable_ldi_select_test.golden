
Welcome to the hypertable command interpreter.
For information about Hypertable, visit http://www.hypertable.org/

Type 'help' for a list of commands, or 'help shell' for a
list of shell meta commands.

USE '/test';
DROP TABLE IF EXISTS hypertable;
CREATE TABLE hypertable(TestColumnFamily);
LOAD DATA INFILE ROW_KEY_COLUMN=rowkey "-" INTO TABLE hypertable;

Welcome to the hypertable command interpreter.
For information about Hypertable, visit http://www.hypertable.org/

Type 'help' for a list of commands, or 'help shell' for a
list of shell meta commands.

USE 'test';
SELECT * FROM hypertable;
key1	TestColumnFamily	tcf1
key2	TestColumnFamily	tcf2
DROP TABLE IF EXISTS hypertable;
#row	column	value
key1	TestColumnFamily	tcf1
key2	TestColumnFamily	tcf2
